core/blockchain.go

```diff
@@ -1272,6 +1287,156 @@ func (bc *BlockChain) WriteBlockWithState(
 	return CanonStatTy, nil
 }
 
+func (bc *BlockChain) pruneBeaconChain(db ethdb.Database, maxBlockNum uint64, maxEpoch *big.Int) {
+	// check is start
+	if atomic.CompareAndSwapInt32(&bc.isPruneBeaconChainRunning, 0, 1) {
+		defer atomic.StoreInt32(&bc.isPruneBeaconChainRunning, 0)
+
+		// limit time spent
+		startTime := time.Now()
+		deletedBlockCount := 0
+		deletedValidatorSnapshot := 0
+		deleter := db.NewBatch()
+		deleteRange := make([][2][]byte, 0)
+
+		defer func() {
+			if err := deleter.Write(); err != nil {
```

Comment:
Looks like the deleter batch is a little too overwhelmed. Have you done metric on memory usage?

Reply:
leveldb is very good for this optimization, but I also added a memory limit now, it will not use more than 4MB of memory